### PR TITLE
[RFC] Proposal to merge the Patatrack development into CMSSW 10.5.x

### PR DESCRIPTION
**Pull request by fwyzard**
Note that the development of the Patatrack branch dates back at least to 10.1.x. Hopefully our development branch has been kept enough in sync with the master branch to avoid big conflicts.

Currently, not all the developments we would like to integrate have been merged in the development branch: see https://github.com/cms-patatrack/cmssw/issues/200 for the status of the Patatrack branch with respect to readiness for merging upstream.

### CODE REVIEW

**Comment by civanch** [comment]
@fwyzard , is it an optimal approach that we have a mix of ordinary c++ and cuda code inside one library? For example there are few classes added to SimTracker/TrackerHitAssociation, may be a new library can be created under SimTracker?

**Comment by fwyzard** [comment]
See https://github.com/cms-sw/cms-bot/pull/1064 for the proposal of a new "heterogeneous" category for the new packages.

**Comment by fwyzard** [comment]
> is it an optimal approach that we have a mix of ordinary c++ and cuda code inside one library?

The answer may depend on what you mean by "cuda code":
  - the code compiled to run on the device (i.e. device functions and kernels) is actually stored in separate ELF sections of the host shared libraries, so CUDA-unaware code does not see it; this is similar to how debug informations are stored;
  - the code compiled to run on the host and interact with the device (i.e. the code to launch the kernels, or code using CUDA library calls) is standard object code; its main impact on the library is the added dependency on the CUDA driver and runtime libraries.

So the question boils down to, is it a problem if we add a dependency on the CUDA libraries to existing plugins ?
I do not think it is, but I'm happy to hear people's opinions.

**Comment by makortel** [comment]
> So the question boils down to, is it a problem if we add a dependency on the CUDA libraries to existing plugins ?

Let's extend the question to data formats as well.

The transient products passed between CUDA EDProducers are relatively simple, but do depend on the CUDA runtime library because of the memory allocator. Here the choice was to **not** add dependence on the CUDA libraries to `DataFormats`, but to create a new "tree" for CUDA-specific data formats (mostly driven by caution). If it turns out that this separation is not needed, there would be some maintenance benefits in placing the transient CUDA data formats among the regular ones.

File: RecoLocalTracker/SiPixelClusterizer/python/SiPixelClusterizer_cfi.py [hunk 1]
```diff
@@ -32,6 +32,10 @@
     Phase2KinkADC = cms.int32(8),
 )
 
+# *only for the cms-patatrack repository*
+# ensure reproducibility for CPU <--> GPU comparisons
+siPixelClusters.payloadType = "HLT"
```
Comment (mmusich) [inline RecoLocalTracker/SiPixelClusterizer/python/SiPixelClusterizer_cfi.py]: is this really meant for CMSSW?

**Comment by davidlange6** [comment]
please test

**Comment by davidlange6** [comment]
Your PR is unmergeable. Please have a look and possibly rebase it.

File: RecoLocalTracker/SiPixelClusterizer/python/SiPixelClusterizer_cfi.py (hunk 1 above)
Comment (fwyzard) [reply RecoLocalTracker/SiPixelClusterizer/python/SiPixelClusterizer_cfi.py]: No, that's one of the changes we need to revert (or maybe make conditional on the GPU-only workflows, for the time being).
Noted in https://github.com/cms-patatrack/cmssw/issues/200 .

**Comment by fwyzard** [comment]
I know, fixing the conflict is on the TODO list :-)

**Comment by vargasa** [comment]
@fwyzard I'm not sure why @cmsbuild merge commits appear on this PR. But make sure you use `rebase` instead of `merge`. Probably you can resolve some of the conflicts by  `cherrypick`ing commits based on `cmssw/master` branch but this can be a little bit tedious as you have more than 250+ of them

**Comment by fwyzard** [comment]
@VinInn 
> The conflict is trivial (one line) ca be solved with the web interface...

Well, yes, but unfortunately doing that would likely break the CMSSW_10_4_X_Patatrack development branch.
We know how to deal with the two (?) conflicts, I will commit a merge in our branch once 10.4.0-pre3 is out.

**Comment by fwyzard** [comment]
@vargasa unfortunately I cannot really rebase or cherry-pick the changes: our development branch has been regularly synchronised with the `master` branch over the past year, and if I were to rebase or cherry-pick now I would have to fix again all the conflicts...

As long as the diff is correct, I wouldn't mind much about the list of commits.

**Comment by makortel** [comment]
I'm not sure we need to override this much :)
```
+  ~HeterogeneousEDProducer() override override override override override = default;
```

**Comment by fwyzard** [comment]
indeed :)

**Comment by fwyzard** [comment]
I'll apply a more reasonable version of the code checks.

**Comment by davidlange6** [comment]
please test

**Comment by fwyzard** [comment]
Wouldn't it be nice if the bot could catch all the code check issues the first time ?

**Comment by davidlange6** [comment]
keep in mind that until a PR merges and compiles, the code checks are not likely to be correct (as they rely on a compiler..) i'm surprised they work at all when the PR does not merge (maybe that a bad thing..)

**Comment by VinInn** [comment]
in this case conflicts are just in python...

**Comment by davidlange6** [comment]
i'm just basing my comment on past experience.. (I'm not sure I've seen a case where the code checks behave strangely that didn't end up with there being a problem with the PR, but maybe this will be such an example)

**Comment by VinInn** [comment]
Does anybody understand why clang-tidy suggests triple override?
```
-   ~PixelCPEFast();
+   ~PixelCPEFast() override override override;
```

**Comment by fwyzard** [comment]
It doesn't - that happens when you run `clang-tidy` in parallel on multiple source files that include that same header file.
`scram build code-checks` uses a more robust approach which collapses the "fixes" and avoids the duplication.

**Comment by andrius-k** [comment]
please test

**Comment by fwyzard** [comment]
How do I tell cmsbuild *not* to run any tests until I say so ?

**Comment by fwyzard** [comment]
hold

**Comment by vargasa** [comment]
> regularly synchronised with the `master` branch over the past year,

Hi @fwyzard. How do you do this synchronization? Just curious :smile_cat: 

> As long as the diff is correct, I wouldn't mind much about the list of commits.

The problem is that checking the diff itself is a pain as right now 'looks like' more than 210 files to check. But don't mind me I'm out of context

**Comment by fwyzard** [comment]
@vargasa 
> How do you do this synchronization?

I `git merge` the CMSSW master into our development branch, and fix the conflicts that arise.
I will do it for CMSSW_10_4_0_pre3 today or during the weekend.

> The problem is that checking the diff itself is a pain as right now 'looks like' more than 210 files to check.

That is because there really are over 200 files added or modified by this PR :-)

**Comment by davidlange6** [comment]
please test

**Comment by vargasa** [comment]
Hi @fwyzard 

> I `git merge` the CMSSW master into our development branch, and fix the conflicts that arise.

For next time, [it is better](https://www.atlassian.com/git/tutorials/merging-vs-rebasing) if you `git rebase masterOfficial` so commits that are already on the master are not shown on the PR and maybe it will be easier to go through the changes, as well as it will take care of the commit history (which is also important for documentation purposes). Just a suggestion :wink:

**Comment by fwyzard** [comment]
@vargasa, for the next time, it would be better to consider the context before giving advice.

The article you so helpfully linked states very nicely that

> Once you understand what rebasing is, the most important thing to learn is when not to do it. The golden rule of git rebase is to never use it on public branches.

This branch has been under development for almost one year, across multiple release cycles, integrating contributions from around ten people: rebasing it and rewriting the git history to follow the CMSSW master branch would have meant - each time - breaking any ongoing development branch they had.

**Comment by davidlange6** [comment]
please test

**Comment by davidlange6** [comment]
please test

**Comment by fwyzard** [comment]
@cmsbuild, please test

**Comment by davidlange6** [comment]
please test

**Comment by fwyzard** [comment]
@cmsbuild, please test

**Comment by slava77** [comment]
@fwyzard @smuzaffar 
do you know what happened with the tests in this PR?
https://github.com/cms-sw/cmssw/pull/25353#issuecomment-447316827

**Comment by smuzaffar** [comment]
enable GPU
@fwyzard , can you please make code-checks happy (by applying the patch https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-25353/7870/git-diff.patch and resolving the conflicts? After that we can trigger the tests.

**Comment by fwyzard** [comment]
@smuzaffar , looks like https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-25353/7870/git-diff.patch has some spurious suggestions:
```diff
diff --cc EventFilter/SiPixelRawToDigi/python/SiPixelRawToDigi_cfi.py
index d56f03a59d9,528ffa2683f..00000000000
--- a/EventFilter/SiPixelRawToDigi/python/SiPixelRawToDigi_cfi.py
+++ b/EventFilter/SiPixelRawToDigi/python/SiPixelRawToDigi_cfi.py
@@@ -20,5 -21,10 +21,15 @@@ siPixelDigis.CablingMapLabel = cms.stri
  
  from Configuration.Eras.Modifier_phase1Pixel_cff import phase1Pixel
  phase1Pixel.toModify(siPixelDigis, UsePhase1=True)
++<<<<<<< HEAD
 +from Configuration.ProcessModifiers.premix_stage2_cff import premix_stage2
 +premix_stage2.toModify(siPixelDigis, BadPixelFEDChannelsInputLabel = "mixData")
++=======
+ 
+ _siPixelDigis_gpu = RecoLocalTracker.SiPixelClusterizer.siPixelDigiHeterogeneousConverter_cfi.siPixelDigiHeterogeneousConverter.clone()
+ _siPixelDigis_gpu.includeErrors = cms.bool(True)
+ 
+ from Configuration.ProcessModifiers.gpu_cff import gpu
+ gpu.toReplaceWith(siPixelDigis, _siPixelDigis_gpu)
+ 
++>>>>>>> cms-sw/refs/pull/25353/head
```

Is that because of the merge conflict ?

**Comment by smuzaffar** [comment]
yes that is due to merge conflicts.

**Comment by fwyzard** [comment]
@smuzaffar can we use #25647 to test the GPU infrastructure ?

**Comment by smuzaffar** [comment]
yes we can, you need to resolve the merge conflict first for EventFilter/SiPixelRawToDigi/python/SiPixelRawToDigi_cfi.py.

**Comment by fwyzard** [comment]
The conflict should be solved there.
I'm waiting for the code checks.

**Comment by fwyzard** [comment]
@cmsbuild, please test

**Comment by fwyzard** [comment]
Please test

**Comment by fwyzard** [comment]
@cmsbuild, please test

**Comment by fwyzard** [comment]
@cmsbuild please test

**Comment by fwyzard** [comment]
Please check https://github.com/cms-sw/cmssw/pull/25647 instead.

**Pull request closed without merging**